ca/certificate-authority.go

```diff
@@ -215,176 +217,397 @@ func (ca *CertificateAuthorityImpl) RevokeCertificate(serial string, reasonCode
-		return emptyCert, err
+		ca.Log.AuditErr(err)
+		return emptyCertRequest, err
 	}
 	if badSignatureAlgorithms[csr.SignatureAlgorithm] {
 		err = core.MalformedRequestError("Invalid signature algorithm in CSR")
 		// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
-		ca.log.AuditErr(err)
-		return emptyCert, err
+		ca.Log.AuditErr(err)
+		return emptyCertRequest, err
 	}
 
+	// Verify that the hostnames in the CSR are ones that we will issue for
 	// Pull hostnames from CSR
 	// Authorization is checked by the RA
-	commonName := ""
-	hostNames := make([]string, len(csr.DNSNames))
-	copy(hostNames, csr.DNSNames)
-	if len(csr.Subject.CommonName) > 0 {
-		commonName = csr.Subject.CommonName
-		hostNames = append(hostNames, csr.Subject.CommonName)
-	} else if len(hostNames) > 0 {
-		commonName = hostNames[0]
-	} else {
+	commonName, hostNames := uniqueHostNames(csr)
+	if len(hostNames) == 0 {
 		err = core.MalformedRequestError("Cannot issue a certificate without a hostname.")
 		// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
-		ca.log.AuditErr(err)
-		return emptyCert, err
+		ca.Log.AuditErr(err)
+		return emptyCertRequest, err
 	}
 
-	// Collapse any duplicate names.  Note that this operation may re-order the names
-	hostNames = core.UniqueNames(hostNames)
+	// Collapse any duplicate names and check that there aren't too many.
+	// Note that this operation may re-order the names.
 	if ca.MaxNames > 0 && len(hostNames) > ca.MaxNames {
 		err = core.MalformedRequestError(fmt.Sprintf("Certificate request has %d > %d names", len(hostNames), ca.MaxNames))
-		ca.log.WarningErr(err)
-		return emptyCert, err
+		ca.Log.WarningErr(err)
+		return emptyCertRequest, err
 	}
 
 	// Verify that names are allowed by policy
 	identifier := core.AcmeIdentifier{Type: core.IdentifierDNS, Value: commonName}
-	if err = ca.PA.WillingToIssue(identifier, regID); err != nil {
+	if err = ca.PA.WillingToIssue(identifier, req.RegistrationID); err != nil {
 		err = core.MalformedRequestError(fmt.Sprintf("Policy forbids issuing for name %s", commonName))
 		// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
-		ca.log.AuditErr(err)
-		return emptyCert, err
+		ca.Log.AuditErr(err)
+		return emptyCertRequest, err
 	}
 	for _, name := range hostNames {
 		identifier = core.AcmeIdentifier{Type: core.IdentifierDNS, Value: name}
-		if err = ca.PA.WillingToIssue(identifier, regID); err != nil {
+		if err = ca.PA.WillingToIssue(identifier, req.RegistrationID); err != nil {
 			err = core.MalformedRequestError(fmt.Sprintf("Policy forbids issuing for name %s", name))
 			// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
-			ca.log.AuditErr(err)
-			return emptyCert, err
+			ca.Log.AuditErr(err)
+			return emptyCertRequest, err
 		}
 	}
 
-	notAfter := ca.Clk.Now().Add(ca.ValidityPeriod)
+	// Fill in the creation time on the request
+	// The expiry time, regID, and CSR are set by the RA on the way in.
```

Comment:
... and likewise, in RA.NewCertificate(), check that the WFE has filled in the proper fields.